proposed/2022/cpm-dotnet-cli-add-package-support.md

```diff
@@ -0,0 +1,134 @@
+```xml
+<Project>
+    <PropertyGroup>
+        <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+    </PropertyGroup>
+    <ItemGroup>
+    </ItemGroup>
+</Project>
+```
+The .csproj file: <br>
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+    <PropertyGroup>
+        <TargetFramework>net6.0</TargetFramework>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>enable</Nullable>
+    </PropertyGroup>
+
+    <ItemGroup>
+    </ItemGroup>
+</Project>
+```
+
+#### After `add package` is executed:<br>
+
+The props file: <br>
+```xml
+<Project>
+    <PropertyGroup>
+        <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+    </PropertyGroup>
+    <ItemGroup>
+    <PackageVersion Include="Newtonsoft.Json" Version="13.0.1"/>
+    </ItemGroup>
+</Project>
+```
+The .csproj file: <br>
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+    <PropertyGroup>
+        <TargetFramework>net6.0</TargetFramework>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>enable</Nullable>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Newtonsoft.Json"/>
+    </ItemGroup>
+</Project>
+```
+
+In case there are multiple `Directory.packages.props` files in the repo, the props file that is closest must be considered.
+
+```xml
+Repository
+ |-- Directory.Packages.props
+ |-- Solution1
+    |-- Directory.Packages.props
+    |-- Project1
+ |-- Solution2
+    |-- Project2
+```
+
+In the above example, the following scenarios are possible:
+1. Project1 will evaluate the Directory.Packages.props file in the Repository\Solution1\ directory.
+2. Project2 will evaluate the Directory.Packages.props file in the Repository\ directory.
+
+***Sourced from https://devblogs.microsoft.com/nuget/introducing-central-package-management/
+
+### 2. The package reference does exist
+If the package already exists in `Directory.packages.props` the version should be updated in `Directory.packages.props`. The package version should either be the latest version or the one specified in the CLI command. The `<PackageReference>` in the project file should not change.
+
+#### Before `add package` is executed:<br>
+```xml
+<Project>
+    <PropertyGroup>
+        <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+    </PropertyGroup>
+    <ItemGroup>
+    <PackageVersion Include="Newtonsoft.Json" Version="12.0.1"/>
+    </ItemGroup>
+</Project>
+```
+
+#### After `add package` is executed:<br>
+```xml
+<Project>
+    <PropertyGroup>
+        <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+    </PropertyGroup>
+    <ItemGroup>
+    <PackageVersion Include="Newtonsoft.Json" Version="13.0.1"/>
+    </ItemGroup>
+</Project>
+```
+
+In case there are multiple `Directory.packages.props` files in the repo, all of them should be searched for a matching package reference. Once a matching package reference is found, its package version can be updated appropriately. 
```

Comment:
I imagine the intention here is if multiple Directory.Packages.props file apply correct? 

This is something that the user can only do manually.

Reply:
> I imagine the intention here is if multiple Directory.Packages.props file apply correct?

You are correct. Incase if others are curious, as per [`Central Package Management rules`](https://devblogs.microsoft.com/nuget/introducing-central-package-management/#central-package-management-rules), only the closest `Directory.Packages.props` file is evaluated for a given project. 
```
Repository
 |-- Directory.Packages.props
 |-- Solution1
     |-- Directory.Packages.props 
     |-- Project1
 |-- Solution2
     |-- Project2
```
However, I think customers can specify an `Import` statement in `Solution1\Directory.Packages.props` file to import properties from `Repository\Directory.Packages.props`

```
Repository
 |-- Directory.Packages.props
 |-- Solution1
     |-- Directory.Packages.props (specifies an import statement pointing to the props file in the parent folder)
     |-- Project1
 |-- Solution2
```

> This is something that the user can only do manually.

Once we get clarity on https://github.com/NuGet/Home/pull/11849/files#r890672222 conversation then we consider this scenario as out of scope / future possibilities for this work.

Reply:
There might be some implementational complexity with multiple files. 
Example, what if both files have it? 
You'd then need to know where it's coming from, which is probably more complicated than a V1. 

I'd encourage explicitly saying that scenarios with multiple Directory.Packages.props are out of scope for now.

Reply:
Discussed about this scenario offline with @jeffkl. Yes, scenarios with multiple `Directory.Packages.props` are out of scope for now. Thanks Nikolche. Created a PR with revised proposal https://github.com/NuGet/Home/pull/11873